src/kernels/composable_kernel/include/kernel_algorithm/reduction_functions.hpp

```diff
@@ -0,0 +1,623 @@
+/*******************************************************************************
+ *
+ * MIT License
+ *
+ * Copyright (c) 2020 Advanced Micro Devices, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+#ifndef CK_REDUCTION_FUNCTIONS_HPP
+#define CK_REDUCTION_FUNCTIONS_HPP
+
+#include <config.hpp>
+
+#include "reduction_common.hpp"
+#include "reduction_operator.hpp"
+
+namespace ck {
+namespace detail {
+
+template <typename T>
+__device__ bool IsNan(T x)
+{
+    // for float and double, use the builtin hip kernel functions
+    return (isnan(x));
+};
+
+template <>
+__device__ bool IsNan<half>(half x)
```

Comment:
Instead of use ```half```, please use ```half_t``` defined in ```composable_kernel/include/utility/float_type.hpp```

The reason is compiler team recommed to use ```_fp16``` instead of native ```half```, so we created a typedef for fp16 type

Reply:
The "half_t" is "_Float16",  which seems be a compiler built-in type.  Can I use math functions like "isnan" with "_Floa616" ?

Reply:
```_Float16``` should not be used anyway except in ```float_type.hpp```, so if ```isnan``` needs to use _Float16 directly, you need to use it inside ```float_type.hpp```

BTW ````half``` should not be used anywhere.

Reply:
What I mean is  whether _Float16/half  can be used with those HIP built-in functions  isnan(), isinf() ... .

Reply:
What I mean is wether  _Float16/half_t can be used with the HIP built-in functions like isnan(), isinf() ...

Reply:
Replaced half by half_t .  https://github.com/ROCmSoftwarePlatform/MIOpen/pull/322/commits/c531624a8763e9af812c2255a8382983d9ea76bc